Bug 1536687 - Add asb addon updates

eriknelson opened the pull request:
* Pull templates from the upstream repo to ensure things stay up to
date.
* Added an additional template that will permission the developer user
correctly so that it can support apb cli functionality.
* Removed unecessary templates/
* Added some docs re: apb cli support

~~**Shouldn't merge before the `apb-1.1.6-1` release and tag are available from the [apb tooling repo](https://github.com/ansibleplaybookbundle/ansible-playbook-bundle)**. We're expecting this to happen tomorrow, 01/22/18.~~ apb-1.1.6-1 has been released.

**NOTE:** Requires the [latest minishift](http://artifacts.ci.centos.org/minishift/minishift/master/474/) binary for testing to pick up the fix for the bug reported here: https://github.com/minishift/minishift/issues/1698

Closes #81

djzager commented on add-ons/ansible-service-broker/ansible-service-broker.addon:
```diff
@@ -5,8 +5,9 @@
-# Add template to Project
-oc create -n ansible-service-broker -f templates/deploy-ansible-service-broker.template.yaml
+# Add upstream broker deployment template to the project
+oc create -n ansible-service-broker -f https://raw.githubusercontent.com/openshift/ansible-service-broker/master/templates/deploy-ansible-service-broker.template.yaml
```
Just a note, the template on the master branch will only work with origin versions `>= 3.7` (because of `Cluster*` from the Service Catalog).

eriknelson replied:
That's correct, thanks @djzager. It would actually be possible to parametrize that resource name if pre 3.7 support is desired.

djzager commented on add-ons/ansible-service-broker/ansible-service-broker.addon:
```diff
@@ -1,13 +1,14 @@
 # Description: Deploys the Ansible Service Broker
-# Required-vars:
+# Required-vars: BROKER_REPO_TAG, APBTOOLS_REPO_TAG, DOCKERHUB_ORG
+# Var-Defaults: BROKER_REPO_TAG=ansible-service-broker-1.1.6-1, APBTOOLS_REPO_TAG=apb-1.1.6-1, DOCKERHUB_ORG=ansibleservicebroker
```
ansibleplaybookbundle?

eriknelson replied:
Nice catch! Thank you.

eriknelson commented:
It can be tested with `--addon-env APBTOOLS_REPO_TAG=master` before an apply, to override the future tag.

eriknelson commented:
apb-1.1.6-1 has been released, this is ready for merge and should work with addon-env defaults.

gbraad commented:
@minishift/minishift-dev WDYT?

LalatenduMohanty commented:
@sabre1041 can you please take a look, since you contributed the initial code base .

LalatenduMohanty commented on add-ons/ansible-service-broker/ansible-service-broker.addon:
```diff
@@ -1,12 +1,14 @@
 # Description: Deploys the Ansible Service Broker
-# Required-vars:
+# Required-vars: BROKER_REPO_TAG, APBTOOLS_REPO_TAG, DOCKERHUB_ORG
+# Var-Defaults: BROKER_REPO_TAG=ansible-service-broker-1.1.8-1, APBTOOLS_REPO_TAG=apb-1.1.6-1, DOCKERHUB_ORG=ansibleplaybookbundle
```
We must define OpenShift-Version metadata as >=3.7.1. Refer https://docs.openshift.org/latest/minishift/using/addons.html#addon-openshift-version-semantics

LalatenduMohanty commented:
@eriknelson The PR is failing for me.  Also I have created an issue  #81 for your PR. Please squash your commits and update mention the issue number in your commit for easier tracking of the PR. 
```
$ minishift addon apply ansible-service-broker
-- Applying addon 'ansible-service-broker':........Error applying the add-on: Error executing command 'ssh command error:
command : docker exec origin /bin/bash -c $'oc process ansible-service-broker -n ansible-service-broker -p DOCKERHUB_ORG=ansibleplaybookbundle -p ROUTING_SUFFIX=192.168.42.55.nip.io -p ETCD_TRUSTED_CA_FILE=/var/run/etcd-auth-secret/ca.crt -p BROKER_CLIENT_CERT_PATH=/var/run/asb-etcd-auth/client.crt -p BROKER_CLIENT_KEY_PATH=/var/run/asb-etcd-auth/client.key -p ETCD_TRUSTED_CA="$(base64 /tmp/etcd-cert/cert.pem)" -p BROKER_CLIENT_CERT="$(base64 /tmp/etcd-cert/MyClient1.pem)" -p BROKER_CLIENT_KEY="$(base64 /tmp/etcd-cert/MyClient1.key)" -p BROKER_CA_CERT=$(oc get secret -n kube-service-catalog -o go-template=\'{{ range .items }}{{ if eq .type \"kubernetes.io/service-account-token\" }}{{ index .data \"service-ca.crt\" }}{{end}}{{\"\\n\"}}{{end}}\' | tail -n 1) | oc create -n ansible-service-broker -f-'
err     : exit status 1
output  : service "asb" created
service "asb-etcd" created
serviceaccount "asb" created
clusterrolebinding "asb" created
clusterrole "asb-auth" created
clusterrolebinding "asb-auth-bind" created
clusterrole "access-asb-role" created
persistentvolumeclaim "etcd" created
deploymentconfig "asb" created
deploymentconfig "asb-etcd" created
secret "asb-auth-secret" created
secret "registry-auth-secret" created
secret "etcd-auth-secret" created
secret "broker-etcd-auth-secret" created
configmap "broker-config" created
serviceaccount "ansibleservicebroker-client" created
clusterrolebinding "ansibleservicebroker-client" created
secret "ansibleservicebroker-client" created
route "asb-1338" created
error: unable to recognize servicecatalog.k8s.io/v1beta1, Kind=ClusterServiceBroker: no matches for servicecatalog.k8s.io/, Kind=ClusterServiceBroker
':
```
```
[lmohanty@LalatenduM-laptop minishift-addons]$ minishift status
Minishift:  Running
Profile:    minishift
OpenShift:  Running (openshift v3.7.1+ab0f056)
DiskUsage:  16% of 18G
```

budhrg commented on add-ons/ansible-service-broker/ansible-service-broker.addon:
```diff
@@ -1,12 +1,14 @@
 # Name: ansible-service-broker
 # Description: Deploys the Ansible Service Broker
-# Required-vars:
+# Required-vars: BROKER_REPO_TAG, APBTOOLS_REPO_TAG, DOCKERHUB_ORG
```
Typo here, It should be `Required-Vars`. See https://docs.openshift.org/latest/minishift/using/addons.html#addon-dynamic-variables

budhrg commented on add-ons/ansible-service-broker/README.md:
```diff
@@ -1,71 +1,84 @@
-## Prerequisites
+# Prerequisites
+
+This addon is designed for an RBAC enabled Openshift cluster of `>=3.7`. Minishift can be configured to
```
`This addon is designed for a RBAC enabled OpenShift cluster of ...`

djzager replied:
It feels silly defending the original grammar, but here I go. In this scenario RBAC is an initialism, making the original correct.

eriknelson replied:
I'll just pretend like I knew what I was doing :)

budhrg commented on add-ons/ansible-service-broker/README.md:
````diff
@@ -1,71 +1,84 @@
+$ minishift config set openshift-version v3.7.1
+```
 
 Minishift/CDK must be deployed with the Service Catalog enabled
````
Should we specify `CDK` here ? Not sure. /cc @LalatenduMohanty

LalatenduMohanty replied:
@budhrg I think it is fine to mention as it is intended for both. But the minimum criteria for any add-on in this repository is that it would work with upstream `minishift` repository.

budhrg commented on add-ons/ansible-service-broker/README.md:
```diff
@@ -1,71 +1,84 @@
 
-## Deploy the Ansible Service Broker
+If you are planning on building or pushing to the minishift registry, be sure
+to configure your shell using `eval $(minishift docker-env)`. This ensures the
```
This line `This ensures the apb tooling can interface with your cluster's local registry` is bit confusing.

eriknelson commented:
@LalatenduMohanty @budhrg Updated based on feedback.

budhrg commented on add-ons/ansible-service-broker/README.md:
```diff
@@ -1,71 +1,96 @@
+======================
 
+This addon installs the [Ansible Service Broker](https://github.com/openshift/ansible-service-broker). The Ansible Service Broker
+is an Open Service Broker designed to deploy [Ansible Playbook Bundles](https://github.com/ansibleplaybookbundle/ansible-playbook-bundle/) (hereafter referred to as APBs).
```
This will work. :+1:

budhrg commented on add-ons/ansible-service-broker/README.md:
```diff
@@ -1,71 +1,96 @@
+# Prerequisites
 
-## Prerequisites
+This addon is designed for an RBAC enabled Openshift cluster of `>=3.7`. Minishift can be configured to
```
There is typo in `OpenShift` here.

budhrg commented on add-ons/ansible-service-broker/README.md:
````diff
@@ -1,71 +1,96 @@
+$ minishift config set iso-url centos
+```
+
+> centos is required for Fedora, CentOS, or RHEL hosts due to a [known issue](https://docs.openshift.org/latest/minishift/troubleshooting/troubleshooting-misc.html#authentication-required-to-push-image) between
````
This seems to me like a note.

Could you write it as:

```
Note: CentOS ISO is required for Fedora, CentOS, or RHEL hosts due to a [known issue](https://docs.openshift.org/latest/minishift/troubleshooting/troubleshooting-misc.html#authentication-required-to-push-image) between the host's docker client and the default boot2docker iso.
```

LalatenduMohanty commented on add-ons/ansible-service-broker/ansible-service-broker.addon:
```diff
@@ -16,6 +19,8 @@ docker exec origin /bin/bash -c $'yum install -y openssl'
+docker exec origin /bin/bash -c $'oc process apbtools-permission -n ansible-service-broker | oc create -f-'
 
 echo Ansible Service Broker Deployed
+echo User [ developer ] has been permissioned for usage with the apb tools
```
s/permissioned /given permission/

LalatenduMohanty commented:
I am trying to test the add-on but seeing the following issue. 
```
$ minishift addon apply ansible-service-broker
-- Applying addon 'ansible-service-broker':.........
Ansible Service Broker Deployed
User [ developer ] has been permissioned for usage with the apb tools

$ minishift addon remove ansible-service-broker
-- Removing addon 'ansible-service-broker':......the server doesn't have a resource type "broker"
Error removing the add-on: Error executing command 'oc delete broker ansible-service-broker'.

$ minishift status
Minishift:  Running
Profile:    minishift
OpenShift:  Running (openshift v3.7.1+ab0f056)
DiskUsage:  15% of 18G

```

djzager commented on add-ons/ansible-service-broker/ansible-service-broker.addon.remove:
```diff
@@ -10,5 +10,15 @@ oc delete clusterrole asb-auth
 oc delete clusterrole access-asb-role
 oc delete broker ansible-service-broker
```
You could update this line to something like:

```
$ oc delete -n ansible-service-broker -f https://raw.githubusercontent.com/ansibleplaybookbundle/ansible-playbook-bundle/#{APBTOOLS_REPO_TAG}/templates/openshift-permissions.template.yaml
```

This should allow you to remove everything created by the template.

LalatenduMohanty commented:
@eriknelson We are pretty close now. There is one issue i.e. if I run apply -> remove -> again apply , it fails as you do not delete the project i.e. `oc delete project ansible-service-broker`

```
$ minishift addon apply ansible-service-broker
-- Applying addon 'ansible-service-broker':.........
Ansible Service Broker Deployed
User [ developer ] has been given permission for usage with the apb tools

$ minishift addon remove ansible-service-broker
-- Removing addon 'ansible-service-broker':..

$ minishift addon apply ansible-service-broker
-- Applying addon 'ansible-service-broker':.Error from server (AlreadyExists): project.project.openshift.io "ansible-service-broker" already exists
Error applying the add-on: Error executing command 'oc new-project ansible-service-broker'.
```

LalatenduMohanty commented:
@eriknelson I tried the add-on again. After I run the remove command I waited till the project is removed from OpenShift console and applied the add-on again and I am seeing below error. I am going to try again to see if it is related to my setup.

```
$ minishift addon apply ansible-service-broker
-- Applying addon 'ansible-service-broker':.........
Ansible Service Broker Deployed
User [ developer ] has been given permission for usage with the apb tools
$ minishift addon remove  ansible-service-broker
-- Removing addon 'ansible-service-broker':..............

$ minishift addon apply ansible-service-broker
-- Applying addon 'ansible-service-broker':........Error applying the add-on: Error executing command 'ssh command error:
command : docker exec origin /bin/bash -c $'oc process ansible-service-broker -n ansible-service-broker -p DOCKERHUB_ORG=ansibleplaybookbundle -p ROUTING_SUFFIX=192.168.64.16.nip.io -p ETCD_TRUSTED_CA_FILE=/var/run/etcd-auth-secret/ca.crt -p BROKER_CLIENT_CERT_PATH=/var/run/asb-etcd-auth/client.crt -p BROKER_CLIENT_KEY_PATH=/var/run/asb-etcd-auth/client.key -p ETCD_TRUSTED_CA="$(base64 /tmp/etcd-cert/cert.pem)" -p BROKER_CLIENT_CERT="$(base64 /tmp/etcd-cert/MyClient1.pem)" -p BROKER_CLIENT_KEY="$(base64 /tmp/etcd-cert/MyClient1.key)" -p BROKER_CA_CERT=$(oc get secret -n kube-service-catalog -o go-template=\'{{ range .items }}{{ if eq .type \"kubernetes.io/service-account-token\" }}{{ index .data \"service-ca.crt\" }}{{end}}{{\"\\n\"}}{{end}}\' | tail -n 1) | oc create -n ansible-service-broker -f-'
err     : exit status 1
output  : service "asb" created
service "asb-etcd" created
serviceaccount "asb" created
clusterrolebinding "asb" created
clusterrole "asb-auth" created
clusterrolebinding "asb-auth-bind" created
clusterrole "access-asb-role" created
persistentvolumeclaim "etcd" created
deploymentconfig "asb" created
deploymentconfig "asb-etcd" created
secret "asb-auth-secret" created
secret "registry-auth-secret" created
secret "etcd-auth-secret" created
secret "broker-etcd-auth-secret" created
configmap "broker-config" created
serviceaccount "ansibleservicebroker-client" created
clusterrolebinding "ansibleservicebroker-client" created
secret "ansibleservicebroker-client" created
route "asb-1338" created
Error from server (AlreadyExists): object is being deleted: clusterservicebrokers.servicecatalog.k8s.io "ansible-service-broker" already exists
':
```

eriknelson commented:
@LalatenduMohanty It seems the `clusterservicebroker` resource takes a bit to be deleted as well.

The problem here is that a lot of resources are required, and when removing them, you need to wait for them all to be reconciled and cleaned before running another apply. Can you try again and wait for the second apply until that `clusterservicebroker` resource is also cleaned up? If I wait for them to be fully cleaned, the second apply succeeds.

Also, is applying, removing, and applying in a short time frame a realistic use-case?

eriknelson commented:
@LalatenduMohanty @budhrg Can you guys take another look at this? There's a bit of pressure on this issue because the current addon in master here is broken, and we have a bunch of users trying to use it and failing.

Either we need to delete the current ansible-service-broker addon from master, or get this merged.

budhrg commented:
@eriknelson  One last change, update the commit message as 
```
Issue #81 Update ASB add-on for OpenShift 3.7.1
```

budhrg commented:
Merged @eriknelson after amending the commit message :) Thanks for contributing.
